src/appleseed/renderer/kernel/lighting/pathtracer.h

```diff
@@ -452,140 +464,185 @@ size_t PathTracer<PathVisitor, Adjoint>::trace(
-
-        // Update path throughput.
-        if (bsdf_sample.m_probability != BSDF::DiracDelta)
-            bsdf_sample.m_value /= bsdf_sample.m_probability;
-        vertex.m_throughput *= bsdf_sample.m_value;
-
-        // Update bounce counters.
-        ++vertex.m_path_length;
-        diffuse_bounces += (bsdf_sample.m_mode >> ScatteringMode::DiffuseBitShift) & 1;
-        glossy_bounces += (bsdf_sample.m_mode >> ScatteringMode::GlossyBitShift) & 1;
-        specular_bounces += (bsdf_sample.m_mode >> ScatteringMode::SpecularBitShift) & 1;
-
-        // Construct the scattered ray.
-        const foundation::Vector3d incoming(bsdf_sample.m_incoming.get_value());
-        ShadingRay next_ray(
-            vertex.m_shading_point->get_biased_point(incoming),
-            incoming,
-            ray.m_time,
-            ScatteringMode::get_vis_flags(bsdf_sample.m_mode),
-            ray.m_depth + 1);
-        next_ray.m_dir = foundation::improve_normalization<2>(next_ray.m_dir);
-
-        // Compute scattered ray differentials.
-        if (bsdf_sample.m_incoming.has_derivatives())
+            // Path survived the checks.
+            terminate_path = false;
+        }
+        else
         {
-            next_ray.m_rx.m_org = next_ray.m_org + vertex.m_shading_point->get_dpdx();
-            next_ray.m_ry.m_org = next_ray.m_org + vertex.m_shading_point->get_dpdy();
-            next_ray.m_rx.m_dir = next_ray.m_dir + foundation::Vector3d(bsdf_sample.m_incoming.get_dx());
-            next_ray.m_ry.m_dir = next_ray.m_dir + foundation::Vector3d(bsdf_sample.m_incoming.get_dy());
-            next_ray.m_has_differentials = true;
+            //
+            // No bounce.
+            //
+
+            // If there is no BSDF, just continue the current ray without increasing its depth.
+            next_ray = ShadingRay(
+                vertex.get_point(),
+                ray.m_dir,
+                ray.m_time,
+                ray.m_flags,
+                ray.m_depth);
+
+            // Path survived the checks.
+            terminate_path = false;
         }
 
+        if (terminate_path) break;
+
         // Build the medium list of the scattered ray.
         const foundation::Vector3d& geometric_normal = vertex.get_geometric_normal();
         const bool crossing_interface =
             foundation::dot(vertex.m_outgoing.get_value(), geometric_normal) *
             foundation::dot(next_ray.m_dir, geometric_normal) < 0.0;
-        if (vertex.m_bsdf != 0 && crossing_interface)
+        if (crossing_interface)
         {
-            // Refracted ray: inherit the medium list of the parent ray and add/remove the current medium.
+            // Ray goes under the surface:
+            // inherit the medium list of the parent ray and add/remove the current medium.
             if (entering)
             {
-                const float ior =
+                const float ior = !vertex.m_bsdf ? 1.0f :
```

Comment:
We try to use boolean tests (`!x`) for boolean expressions only (i.e. when `x` is of type `bool`) and prefer explicit tests against 0 for other types. Here: `vertex.m_bsdf == nullptr`